### PR TITLE
Provers: split functions between prove_* and check_* versions

### PR DESCRIPTION
**Pull request by lthls**
The prove_* versions will return a Proof when the inputs type strictly
matches the property in question, while the check_* functions will return
a Known_result when the corresponding operation has a result that can be
easily extracted.

As an example, for untagging the `prove_tagging_of_simple` version will
return Unknown on option-like variants, while the `check_tagging_of_simple`
version will assume the block part of the type is irrelevant.

### CODE REVIEW

**Comment by lthls** [comment]
Note: this PR currently contains all the core changes, but could benefit from more documentation and probably a bit of refactoring.

But these issues are not as complex as just getting the code right, and I'm assuming that they can be addressed during the review.

File: middle_end/flambda2/simplify/inlining/call_site_inlining_decision.ml [hunk 1]
```diff
@@ -169,10 +169,10 @@ let might_inline dacc ~apply ~code_or_metadata ~function_type ~simplify_expr
 
 let get_rec_info dacc ~function_type =
   let rec_info = FT.rec_info function_type in
-  match Flambda2_types.prove_rec_info (DA.typing_env dacc) rec_info with
-  | Proved rec_info -> rec_info
+  match Flambda2_types.check_rec_info (DA.typing_env dacc) rec_info with
+  | Known_result rec_info -> rec_info
   | Unknown -> Rec_info_expr.unknown
-  | Invalid -> Rec_info_expr.do_not_inline
+  | Invalid -> (* CR vlaviron: ? *) Rec_info_expr.do_not_inline
```
Comment (lukemaurer) [inline middle_end/flambda2/simplify/inlining/call_site_inlining_decision.ml:175]: I think I wrote this line, the idea being that if something has gone wrong with the rec info, we want to suppress inlining completely. Possibly it should simply be a fatal error, though. When would we expect to see an `Invalid` here?

Comment (lthls) [reply middle_end/flambda2/simplify/inlining/call_site_inlining_decision.ml:175]: From my point of view, `Invalid` means that we are in a situation which can never happen at runtime. Typically, we're in a dead branch for unrelated reasons. So usually (if you look at other uses of the `prove_*` or `check_*` functions), either the `Invalid` case is handled as the `Unknown` case (after all, this is safe), or `Invalid` is propagated further. The reason for the CR is that the code doesn't fit these patterns, and I wanted to know why.

**Comment by lthls** [comment]
Merged in #726

**Pull request closed without merging**